manager/allocator/network/port/port.go

```diff
@@ -0,0 +1,470 @@
+// of PortConfigs which should be written directly without modification to the
+// the Endpoint.Ports field. Nothing should be modified about Endpoint.Ports
+// outside of the port allocator. If this rule is followed, the the port
+// allocator doesn't have to check for cases like two services having the same
+// port allocated; we created all of the fields, and we can know they are
+// consistent.
+type allocator struct {
+	// ports maps the ports in use. essentially, using map as a set
+	ports map[port]struct{}
+}
+
+// port is the minimal representation for the Allocator of a single port.
+// being composed of 2 numeric types, this type is comparable and can be used
+// as a map key
+type port struct {
+	// protocol represented by this port space
+	protocol api.PortConfig_Protocol
+	port     uint32
+}
+
+// String is a quick implementation of the Stringer interface for the port
+// object so that we can pass it to string format calls with just %v
+func (p port) String() string {
+	return fmt.Sprintf("%v/%v", p.port, p.protocol)
+}
+
+// Proposal is the return value of Allocate and DeallocateEndpoint,
+// which contains the data needed to affirmatively alter the state of the
+// Allocator. It exists so that Allocator changes can be abandoned
+// without complex rollback logic. To commit changes to the Allocator, call
+// the proposal's Commit function
+type Proposal interface {
+	// Ports returns the port assignments generated by this proposal
+	Ports() []*api.PortConfig
+
+	// Commit commits the result of the Allocator transaction and alters
+	// the state of the port allocator. Part of the contract of the
+	// Allocator is that it should always return a valid proposal, and so
+	// Commit will always succeed and does not need to return an error.
+	Commit()
+
+	// IsNoop returns true if the Proposal doesn't actually modify the net
+	// allocator state. IsNoop does not mean the ports for an endpoint haven't
+	// changed. It only means that the publish ports marked in use by the
+	// allocator haven't changed.
+	IsNoop() bool
+}
+
+type proposal struct {
+	pa         *allocator
+	ports      []*api.PortConfig
+	allocate   map[port]struct{}
+	deallocate map[port]struct{}
+}
+
+func (prop *proposal) Ports() []*api.PortConfig {
+	if prop.ports == nil {
+		return []*api.PortConfig{}
+	}
+	return prop.ports
+}
+
+// Commit commits the proposal to the port allocator.
+func (prop *proposal) Commit() {
+	if prop.IsNoop() {
+		// nothing to do if proposal is noop, short circuit a bit
+		return
+	}
+	// The pattern here is we're going to free every port in p.deallocate and
+	// then allocate every port in allocate. any over lap results in no net
+	// change
+	for p := range prop.deallocate {
+		delete(prop.pa.ports, p)
+	}
+	for p := range prop.allocate {
+		prop.pa.ports[p] = struct{}{}
+	}
+}
+
+// IsNoop returns true if the ports in use before this proposal are the same as
+// the ports in use after.
+func (prop *proposal) IsNoop() bool {
```

Comment:
this is almost entirely for the benefit of the tests, and i think the "correct" answer may actually be to change the tests to not use the `port_test` package and instead allow them to look at an unexported method or field providing this functionality.

Reply:
I made a PR to your forked repository with a couple optimizations to the `Allocate` loop. The second one is based on this comment from @aaronlehmann.

P.S: If the optimizations get accepted, `IsNoop` results in checking that both `allocate` and `deallocate` are empty, so it may not even need a method if you opt to not use the `port_test` package.